heron/scheduler-core/src/java/com/twitter/heron/scheduler/RuntimeManagerMain.java

```diff
@@ -403,46 +402,69 @@ public void manageTopology()
+   * which could be caused by not fully successful SUBMIT or KILL command. In this
+   * case, we need to skip the validation and allow KILL command to go through.
+   * In case execution state data is available, environment check will be done anyway.
+   * @return true if the topology execution data is found, false otherwise.
+   */
+  protected boolean validateRuntimeManage(
       SchedulerStateManagerAdaptor adaptor,
       String topologyName) throws TopologyRuntimeManagementException {
     // Check whether the topology has already been running
     Boolean isTopologyRunning = adaptor.isTopologyRunning(topologyName);
-
-    if (isTopologyRunning == null || isTopologyRunning.equals(Boolean.FALSE)) {
-      throw new TopologyRuntimeManagementException(
-          String.format("Topology '%s' does not exist", topologyName));
+    boolean hasExecutionData = isTopologyRunning != null && isTopologyRunning.equals(Boolean.TRUE);
+    if (!hasExecutionData) {
+      if (command == Command.KILL) {
+        LOG.warning(String.format("Topology '%s' is not found or not running", topologyName));
+      } else {
+        throw new TopologyRuntimeManagementException(
+            String.format("Topology '%s' does not exist", topologyName));
+      }
     }
 
-    // Check whether cluster/role/environ matched
+    // Check whether cluster/role/environ matched if execution state data is available.
     ExecutionEnvironment.ExecutionState executionState = adaptor.getExecutionState(topologyName);
     if (executionState == null) {
-      throw new TopologyRuntimeManagementException(
-          String.format("Failed to get execution state for topology %s", topologyName));
-    }
-
-    String stateCluster = executionState.getCluster();
-    String stateRole = executionState.getRole();
-    String stateEnv = executionState.getEnviron();
-    String configCluster = Context.cluster(config);
-    String configRole = Context.role(config);
-    String configEnv = Context.environ(config);
-    if (!stateCluster.equals(configCluster)
-        || !stateRole.equals(configRole)
-        || !stateEnv.equals(configEnv)) {
-      String currentState = String.format("%s/%s/%s", stateCluster, stateRole, stateEnv);
-      String configState = String.format("%s/%s/%s", configCluster, configRole, configEnv);
-      throw new TopologyRuntimeManagementException(String.format(
-          "cluster/role/environ does not match. Topology '%s' is running at %s, not %s",
-          topologyName, currentState, configState));
+      if (command == Command.KILL) {
+        LOG.warning(String.format("Topology execution state for '%s' is not found", topologyName));
+      } else {
+        throw new TopologyRuntimeManagementException(
+            String.format("Failed to get execution state for topology %s", topologyName));
+      }
+    } else {
+      // Execution state is available, validate configurations.
```

Comment:
@nwangtw - can you make this as a function?

Reply:
sounds good. let me do it.